pegen/grammar_visualizer.py

```diff
@@ -1,82 +1,114 @@
         if (name := getattr(node, "name", None)):
             return node.__class__.__name__ + f"({node.name})"
         elif (value := getattr(node, "value", None)):
             return node.__class__.__name__ + f"({node.value})"
         else:
             return node.__class__.__name__
 
+    @functools.lru_cache
+    def _total_number_of_children(self, node):
+        """Get the total number of children nodes under a given one recursively"""
+        return sum(self._total_number_of_children(child) for child in self.children(node)) + 1
 
-    def print_grammar(self, current_node, indent='', last=Direction.NEUTRAL, printer=print):
+    def _segment_children(self, node):
+        """Separate children of a given node in two groups so every group has roughly
+        the same number of nodes under it."""
 
-        child_sizes = {child: self.nb_children(child) for child in self.children(current_node)}
+        child_sizes = {child: self._total_number_of_children(child)
+                       for child in self.children(node)}
 
         up = sorted(child_sizes, key=lambda key: child_sizes[key])
 
         down = []
         while up and sum(child_sizes[node] for node in down) < sum(child_sizes[node] for node in up):
             down.append(up.pop())
 
-        for child in up:
-            next_last = Direction.UP if up.index(child) == 0 else Direction.NONE
-            next_indent = (f"{indent}{' ' if last == Direction.UP else '│'}"
-                           f"{' ' * len(self.name(current_node))}")
-            self.print_grammar(child, next_indent, next_last)
-
-        if last == Direction.UP:
-            start_shape = '┌'
-        elif last == Direction.DOWN:
-            start_shape = '└'
-        elif last == Direction.NEUTRAL:
-            start_shape = ' '
-        else:
-            start_shape = '├'
+        return up, down
 
-        if up:
-            end_shape = '┤'
-        elif down:
-            end_shape = '┐'
-        else:
-            end_shape = ''
+    def _next_indent(self, current_indent, n_spaces, direction, last_direction):
+        return f"{current_indent}{' ' if last_direction == direction else '│'}{' ' * n_spaces}"
+
+    def _visit_node_group(self, n_spaces, group, direction, boundary, last_direction, indent, printer):
+        """Visit a group of children recursively printing them accordingly taking into account if they
+           must be printed over the parent node or below it (indicated by *direction*)."""
+
+        for child in group:
+            next_last = direction if child is boundary else Direction.NONE
+            next_indent = self._next_indent(indent, n_spaces, direction, last_direction)
+            self._print_nodes_recursively(child, next_indent, next_last, printer)
 
-        printer(f'{indent}{start_shape}{self.name(current_node)}{end_shape}')
+        # Add an extra line between nodes when we branch out so they don't appear super packed.
```

Comment:
There is an anomaly caused by this part. Look at this output:
```
│     │                  ┌Alt──NamedItem──NameLeaf(list)
│     │                  │
│     │                  ├Alt──NamedItem──NameLeaf(tuple)
│     │                  │
│     │                  ├Alt──NamedItem──NameLeaf(group)
│     │                  │
│     ├Rule(primary)──Rhs┤
│     │                  ├Alt──NamedItem──NameLeaf(NUMBER)
│     │                  │
│     │                  ├Alt──NamedItem──NameLeaf(STRING)
│     │                  │
│     │                  └Alt──NamedItem──NameLeaf(NAME)
```
Note that there's an extra "blank" line before the `Rule(primary)----Rhs` in the middle that shouldn't be there (at least, it breaks the neat visual array of children).